feat: Add reset password

Niharika0104 opened the pull request:
added reset password feature
![image](https://github.com/abhijeetnishal/Password-Shield/assets/124892559/04276b7b-eb15-422a-9c5b-6e78602a136f)

Solves #45

abhijeetnishal commented:
hey @Niharika0104, i guess you have implemented the client side reset password feature right?

Niharika0104 commented:
No I have added backend logic as well

abhijeetnishal commented:
Hey @Niharika0104, backend logic should be in server folder as we have separate backend and frontend it's also necessary for security purposes.

Niharika0104 commented:
@abhijeetnishal I added backend logic in the server folder itself,please check authcontroller  there you can find routes that I have created.

abhijeetnishal commented:
> @abhijeetnishal I added backend logic in the server folder itself,please check authcontroller there you can find routes that I have created.

Ok I will see, don't know it's not showing here in file changes

Niharika0104 commented:
> > @abhijeetnishal I added backend logic in the server folder itself,please check authcontroller there you can find routes that I have created.
> 
> Ok I will see, don't know it's not showing here in file changes 

Sorry my bad I was in client folder when I pushed the changes.I will add those files as in which I added backend logic.

abhijeetnishal commented:
> > > @abhijeetnishal I added backend logic in the server folder itself,please check authcontroller there you can find routes that I have created.
> > 
> > 
> > Ok I will see, don't know it's not showing here in file changes
> 
> @abhijeetnishal Sorry my bad I was in client folder when I pushed the changes.I will add those files as in which I added backend logic.

Ok no problem

Niharika0104 commented:
@abhijeetnishal  now you can check I have pushed all the files

abhijeetnishal commented on server/src/config/dbConnect.ts:
```diff
@@ -1,6 +1,6 @@
 import { Client } from "pg";
 import { Pool } from 'postgres-pool';
-
+import createSchemas from '../models/schemas'
```
Could you explain this line?
why are we importing this?
If we can then it will create a schema on every time database connects right?

Niharika0104 replied:
We can remove this initially I had confusion how are creating schema sorry you can remove that line

abhijeetnishal replied:
Then remove this. You also remove it and update the same PR.

abhijeetnishal commented on server/src/controller/authControllers.ts:
```diff
@@ -155,9 +156,74 @@ const login = async (req: Request, res: Response) => {
   }
 };
 
```
Use Prettier for every file to format the code

abhijeetnishal commented on server/src/controller/authControllers.ts:
```diff
@@ -155,9 +156,74 @@ const login = async (req: Request, res: Response) => {
+  try {
+    const {email}=req.body;
+    
+    const result = await db.pool.query(
```
use JS destructuring

`const { rows } = await db.pool.query(...`

Niharika0104 commented:
Hey should I do all the modifications again

abhijeetnishal commented on server/src/controller/authControllers.ts:
```diff
@@ -155,9 +156,74 @@ const login = async (req: Request, res: Response) => {
+      [email]
+    );
+   
+    if(result.rowCount==0){
```
Now you can check rows[0] or rows.length > 0 condition

Niharika0104 replied:
Any issues if I use that???

abhijeetnishal replied:
There is no issue but it's better to have less and consistent code

abhijeetnishal commented on server/src/controller/authControllers.ts:
```diff
@@ -155,9 +156,74 @@ const login = async (req: Request, res: Response) => {
+    );
+   
+    if(result.rowCount==0){
+      res.json({status:false,error:false});
```
add proper status code and also add message instead of status or error.

Niharika0104 replied:
Okay

abhijeetnishal commented on server/src/controller/authControllers.ts:
```diff
@@ -155,9 +156,74 @@ const login = async (req: Request, res: Response) => {
+    if(result.rowCount==0){
+      res.json({status:false,error:false});
+      console.log(result,"result");
+      return;
```
Directly return like return res.status(200).json...

abhijeetnishal commented on server/src/controller/authControllers.ts:
```diff
@@ -155,9 +156,74 @@ const login = async (req: Request, res: Response) => {
+    let mailOptions = {
+      from:process.env.EMAIL_ADDRESS, // Sender name and address
+      to: email, // Recipient email address
+      subject: 'Password Reset', // Email subject
```
Change subject to Reset password

abhijeetnishal commented on server/src/controller/authControllers.ts:
```diff
@@ -155,9 +156,74 @@ const login = async (req: Request, res: Response) => {
+
+    // Send email
+    await transporter.sendMail(mailOptions);
+    res.json({status:true,error:false})
```
Here also send status code with message

abhijeetnishal commented on server/src/controller/authControllers.ts:
```diff
@@ -155,9 +156,74 @@ const login = async (req: Request, res: Response) => {
+    // Send email
+    await transporter.sendMail(mailOptions);
+    res.json({status:true,error:false})
+    console.log('Password reset email sent successfully.');
```
Remove console.log

abhijeetnishal commented on server/src/controller/authControllers.ts:
```diff
@@ -155,9 +156,74 @@ const login = async (req: Request, res: Response) => {
+    res.json({status:true,error:false})
+    console.log('Password reset email sent successfully.');
+  } catch (error) {
+  res.json({status:false,"error":true})
```
Here also use `return res.status....`

abhijeetnishal commented on server/src/controller/authControllers.ts:
```diff
@@ -155,9 +156,74 @@ const login = async (req: Request, res: Response) => {
+const resetpassword=async (req: Request, res: Response)=>{
+  const { email,password } = req.body;
+      // Execute a SELECT query to check if the email exists
+      try {
```
Here, verify the email exists or not else it will directly reset password.

abhijeetnishal commented on server/src/controller/authControllers.ts:
```diff
@@ -155,9 +156,74 @@ const login = async (req: Request, res: Response) => {
+          [hashedPassword, email]
+        );
+       
+        if (result.rowCount === 0) {
```
Now remove this

abhijeetnishal commented on server/src/controller/authControllers.ts:
```diff
@@ -155,9 +156,74 @@ const login = async (req: Request, res: Response) => {
+        res.status(200).json({ message: "Password updated successfully",status:true,error:false });
+      } catch (error) {
+        console.error("Error updating password: ", error);
+        res.status(500).json({ message: "Internal Server Error",status:false,error:true });
```
Use return

abhijeetnishal commented on server/src/models/schemas.ts:
```diff
@@ -29,6 +29,7 @@ const createSchemas = async(req: Request, res: Response)=> {
             )
             `
         );
+        console.log("helo")
```
remove console.log

abhijeetnishal commented on server/src/routers/authRoutes.ts:
```diff
@@ -15,6 +15,8 @@ userRouter.post("/login", login);
 
 //create an endpoint for logout
 userRouter.post("/logout", logout);
+userRouter.post("/sendemail", sendPasswordResetEmail);
```
You can use middleware also to check user validation.

abhijeetnishal commented:
Hey @Niharika0104, we can refer token validation in development branch as i have updated the backend code you can refer and change accordingly

abhijeetnishal commented on client/src/components/auth/ResetPassword.jsx:
```diff
@@ -0,0 +1,69 @@
+  const [email, setEmail] = useState('');
+  
+  //eslint-disable-next-line
+  const cookies = new Cookies();
```
Why do we need this cookies?

abhijeetnishal commented:
I will review the frontend code later until then work on the reviewed code

abhijeetnishal commented:
Hey @Niharika0104, are you working on it or you want to assign this issue to someone else?

Niharika0104 commented:
@abhijeetnishal I busy with something these days if you are willing to give me 3 days I will push the modified changes soon.

abhijeetnishal commented:
Hey @Niharika0104, if you are busy then inform us as it is almost more than 2 weeks. Other contributers are asking.

Now our codebase has many changes. So resolve the conflicts or just pull or clone and add changes. Don't try to force push it.

abhijeetnishal commented:
Closing this PR as inactive.
